### PR TITLE
[MASSEMBLY-973] Support zstd format

### PR DESCRIPTION
**Pull request by pleeplop**
This pull request provides support for zstd format (provided by plexus-archiver and zstd-jni)
https://github.com/facebook/zstd
https://github.com/luben/zstd-jni

Let me know if you have any feedback.

### CODE REVIEW

File: src/test/java/org/apache/maven/plugins/assembly/archive/DefaultAssemblyArchiverTest.java [hunk 1]
```diff
@@ -467,6 +467,25 @@ public void testCreateTarArchiver_TXzFormat_ShouldInitializeXzCompression()
         verify( archiverManager ).getArchiver( "tar" );
     }
 
+    @Test
+    public void testCreateTarArchiver_TarZstFormat_ShouldInitializeXzCompression()
```
Comment (michael-o) [inline src/test/java/org/apache/maven/plugins/assembly/archive/DefaultAssemblyArchiverTest.java:471]: method name still says XZ

**Review by michael-o — CHANGES_REQUESTED** [review]

File: src/test/java/org/apache/maven/plugins/assembly/archive/DefaultAssemblyArchiverTest.java (hunk 1 above)
Comment (pleeplop) [reply src/test/java/org/apache/maven/plugins/assembly/archive/DefaultAssemblyArchiverTest.java]: It should be fixed now.

**Comment by michael-o** [comment]
I don't understand how this is supposed to run when zstd is not on the classpath?

**Comment by pleeplop** [comment]
> 

zstd-jni should be on classpath because it is a dependency of plexus-archiver.

**Review by michael-o — CHANGES_REQUESTED** [review]
Prepare the code to have a linkage error when the native library is not available. Otherwise the test isn't portable anymore.

**Comment by michael-o** [comment]
Test:
```
mosipov@bsd1srv:/tmp/maven-assembly-plugin (zstd =)
$ zip -d /home/mosipov/.m2/repository/com/github/luben/zstd-jni/1.5.2-5/zstd-jni-1.5.2-5.jar  freebsd/amd64/ freebsd/amd64/libzstd-jni-1.5.2-5.so
deleting: freebsd/amd64/
deleting: freebsd/amd64/libzstd-jni-1.5.2-5.so
```
results in:
```
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  3.160 s
[INFO] Finished at: 2022-11-26T18:08:23+01:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-assembly-plugin:3.5.0-SNAPSHOT:single (default-cli) on project format-test: Execution default-cli of goal org.apache.maven.plugins:maven-assembly-plugin:3.5.0-SNAPSHOT:single failed: An API incompatibility was encountered while executing org.apache.maven.plugins:maven-assembly-plugin:3.5.0-SNAPSHOT:single: java.lang.UnsatisfiedLinkError: no zstd-jni-1.5.2-5 in java.library.path
[ERROR] Unsupported OS/arch, cannot find /freebsd/amd64/libzstd-jni-1.5.2-5.so or load zstd-jni-1.5.2-5 from system libraries. Please try building from source the jar or providing libzstd-jni-1.5.2-5 in your system.
[ERROR] -----------------------------------------------------
[ERROR] realm =    plugin>org.apache.maven.plugins:maven-assembly-plugin:3.5.0-SNAPSHOT
[ERROR] strategy = org.codehaus.plexus.classworlds.strategy.SelfFirstStrategy
[ERROR] urls[0] = file:/tmp/maven-assembly-plugin/target/local-repo/org/apache/maven/plugins/maven-assembly-plugin/3.5.0-SNAPSHOT/maven-assembly-plugin-3.5.0-SNAPSHOT.jar
[ERROR] urls[1] = file:/tmp/maven-assembly-plugin/target/local-repo/org/apache/maven/shared/maven-common-artifact-filters/3.3.1/maven-common-artifact-filters-3.3.1.jar
[ERROR] urls[2] = file:/tmp/maven-assembly-plugin/target/local-repo/org/apache/maven/shared/maven-artifact-transfer/0.13.1/maven-artifact-transfer-0.13.1.jar
[ERROR] urls[3] = file:/tmp/maven-assembly-plugin/target/local-repo/org/sonatype/aether/aether-util/1.7/aether-util-1.7.jar
[ERROR] urls[4] = file:/tmp/maven-assembly-plugin/target/local-repo/org/sonatype/sisu/sisu-inject-bean/1.4.2/sisu-inject-bean-1.4.2.jar
[ERROR] urls[5] = file:/tmp/maven-assembly-plugin/target/local-repo/org/sonatype/sisu/sisu-guice/2.1.7/sisu-guice-2.1.7-noaop.jar
[ERROR] urls[6] = file:/tmp/maven-assembly-plugin/target/local-repo/org/sonatype/plexus/plexus-sec-dispatcher/1.3/plexus-sec-dispatcher-1.3.jar
[ERROR] urls[7] = file:/tmp/maven-assembly-plugin/target/local-repo/org/sonatype/plexus/plexus-cipher/1.4/plexus-cipher-1.4.jar
[ERROR] urls[8] = file:/tmp/maven-assembly-plugin/target/local-repo/org/codehaus/plexus/plexus-component-annotations/2.1.1/plexus-component-annotations-2.1.1.jar
[ERROR] urls[9] = file:/tmp/maven-assembly-plugin/target/local-repo/org/codehaus/plexus/plexus-interpolation/1.26/plexus-interpolation-1.26.jar
[ERROR] urls[10] = file:/tmp/maven-assembly-plugin/target/local-repo/commons-io/commons-io/2.11.0/commons-io-2.11.0.jar
[ERROR] urls[11] = file:/tmp/maven-assembly-plugin/target/local-repo/org/apache/maven/shared/maven-filtering/3.3.0/maven-filtering-3.3.0.jar
[ERROR] urls[12] = file:/tmp/maven-assembly-plugin/target/local-repo/org/sonatype/plexus/plexus-build-api/0.0.7/plexus-build-api-0.0.7.jar
[ERROR] urls[13] = file:/tmp/maven-assembly-plugin/target/local-repo/org/codehaus/plexus/plexus-io/3.4.0/plexus-io-3.4.0.jar
[ERROR] urls[14] = file:/tmp/maven-assembly-plugin/target/local-repo/org/codehaus/plexus/plexus-archiver/4.6.0/plexus-archiver-4.6.0.jar
[ERROR] urls[15] = file:/tmp/maven-assembly-plugin/target/local-repo/org/apache/commons/commons-compress/1.21/commons-compress-1.21.jar
[ERROR] urls[16] = file:/tmp/maven-assembly-plugin/target/local-repo/org/iq80/snappy/snappy/0.4/snappy-0.4.jar
[ERROR] urls[17] = file:/tmp/maven-assembly-plugin/target/local-repo/org/tukaani/xz/1.9/xz-1.9.jar
[ERROR] urls[18] = file:/tmp/maven-assembly-plugin/target/local-repo/com/github/luben/zstd-jni/1.5.2-5/zstd-jni-1.5.2-5.jar
[ERROR] urls[19] = file:/tmp/maven-assembly-plugin/target/local-repo/org/apache/maven/maven-archiver/3.6.0/maven-archiver-3.6.0.jar
[ERROR] urls[20] = file:/tmp/maven-assembly-plugin/target/local-repo/org/codehaus/plexus/plexus-utils/3.4.2/plexus-utils-3.4.2.jar
[ERROR] Number of foreign imports: 1
[ERROR] import: Entry[import  from realm ClassRealm[maven.api, parent: null]]
[ERROR]
[ERROR] -----------------------------------------------------
[ERROR] -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-assembly-plugin:3.5.0-SNAPSHOT:single (default-cli) on project format-test: Execution default-cli of goal org.apache.maven.plugins:maven-assembly-plugin:3.5.0-SNAPSHOT:single failed: An API incompatibility was encountered while executing org.apache.maven.plugins:maven-assembly-plugin:3.5.0-SNAPSHOT:single: java.lang.UnsatisfiedLinkError: no zstd-jni-1.5.2-5 in java.library.path
Unsupported OS/arch, cannot find /freebsd/amd64/libzstd-jni-1.5.2-5.so or load zstd-jni-1.5.2-5 from system libraries. Please try building from source the jar or providing libzstd-jni-1.5.2-5 in your system.
-----------------------------------------------------
realm =    plugin>org.apache.maven.plugins:maven-assembly-plugin:3.5.0-SNAPSHOT
strategy = org.codehaus.plexus.classworlds.strategy.SelfFirstStrategy
urls[0] = file:/tmp/maven-assembly-plugin/target/local-repo/org/apache/maven/plugins/maven-assembly-plugin/3.5.0-SNAPSHOT/maven-assembly-plugin-3.5.0-SNAPSHOT.jar
urls[1] = file:/tmp/maven-assembly-plugin/target/local-repo/org/apache/maven/shared/maven-common-artifact-filters/3.3.1/maven-common-artifact-filters-3.3.1.jar
:
```

**Comment by pleeplop** [comment]
Ok i think i understand your point.
Is the latest change fix the issue?

**Comment by michael-o** [comment]
> Ok i think i understand your point. Does the latest change fix the issue?

The change is incomplete which you handle the exception, you ignore it in the test and don't skip the test:
```
Caused by: org.apache.maven.plugins.assembly.archive.ArchiveCreationException: Unable to find native library for assembly 'format-test'
    at org.apache.maven.plugins.assembly.archive.DefaultAssemblyArchiver.createArchive (DefaultAssemblyArchiver.java:197)
    at org.apache.maven.plugins.assembly.mojos.AbstractAssemblyMojo.execute (AbstractAssemblyMojo.java:523)
    at org.apache.maven.plugins.assembly.mojos.SingleAssemblyMojo.execute (SingleAssemblyMojo.java:60)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
```

Moreover, you might want to map `.tzst` as well:
```
<assembly xsi:schemaLocation="http://maven.apache.org/ASSEMBLY/2.1.1 https://maven.apache.org/xsd/assembly-2.1.1.xsd" xmlns="http://maven.apache.org/ASSEMBLY/2.1.1"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
  <id>format-test</id>
  <formats>
    <format>dir</format>
    <format>zip</format>
    <format>tar</format>
    <format>tar.gz</format>
    <format>tgz</format>
    <format>tar.bz2</format>
    <format>tbz2</format>
    <format>jar</format>
    <format>war</format>
    <format>tar.snappy</format>
    <format>tar.xz</format>
    <format>txz</format>
    <format>tar.zst</format>
  </formats>
```

**Comment by pleeplop** [comment]
> Moreover, you might want to map `.tzst` as well:

Good idea, it should work now.


> The change is incomplete which you handle the exception, you ignore it in the test and don't skip the test:

Ok, I see.
The latest change uses a [selector script](https://maven.apache.org/plugins/maven-invoker-plugin/examples/selector-scripts.html) to skip tests.

Native library available:
```
[INFO] --- maven-invoker-plugin:3.2.2:verify (integration-test) @ maven-assembly-plugin ---
[INFO] -------------------------------------------------
[INFO] Build Summary:
[INFO]   Passed: 146, Failed: 0, Errors: 0, Skipped: 0
[INFO] -------------------------------------------------
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
```

Native library missing:
```
[INFO] --- maven-invoker-plugin:3.2.2:verify (integration-test) @ maven-assembly-plugin ---                                                                                                               [24/1391]
[INFO] -------------------------------------------------                                    
[INFO] Build Summary:                                                                                    
[INFO]   Passed: 145, Failed: 0, Errors: 0, Skipped: 1                                                   
[INFO] -------------------------------------------------                                             
[ERROR] The following builds was skipped:                                                                
[ERROR] *  projects/basic-features/formats-zstd/pom.xml                                                  
[INFO] -------------------------------------------------                                          
[INFO] ------------------------------------------------------------------------   
[INFO] BUILD SUCCESS                                                                                     
[INFO] ------------------------------------------------------------------------
```

**Comment by pleeplop** [comment]
@michael-o Do you want more changes?
Should i squash these commits?

**Comment by michael-o** [comment]
> @michael-o Do you want more changes? Should i squash these commits?

Haven't found the time yet. Will try to complete this month.

**Comment by pleeplop** [comment]
@michael-o Do you find some time to have a look?

**Comment by jira-importer** [comment]
Resolve #1182

**Pull request closed without merging**